### PR TITLE
fix(ErrorReport): Check more thoroughly the sfdx process

### PR DESCRIPTION
**Pull request by louis-bompart**
Before: Just checking if an error was thrown by the process (SFDX sometimes don't even when it fails)
After: Check the error, the stderr pipe and the exit signal.

### CODE REVIEW

**Comment by erocheleau** [comment]
Did you check with long running commands that this was also working?

Like a package install command, you can specify the wait parameter to wait until the package is done installing. Or the metadataapi package deploy, example:

```
sfdx.mdapi.deploy({
    deploydir: DIRECTORY,
    wait: 10,
    targetusername: alias})
.then(...)
.catch(...)
```

File: src/core/commandRunner.ts [hunk 1]
```diff
@@ -25,8 +22,17 @@ export class CommandRunner implements ICommandRunner {
...
-        resolve((buffer as Buffer).toString("utf8"))
+        let cmdResult = spawnSync(fullCommand, actualOptions)
+        if (cmdResult.error) {
+          throw cmdResult.error
+        }
+        if (cmdResult.stderr) {
+          throw cmdResult.stderr.toString("utf8")
+        }
+        if (!cmdResult.signal) {
+          throw new Error(`No error specified, exit signal=${cmdResult.signal}`)
```
Comment (erocheleau) [inline src/core/commandRunner.ts]: Doesn't work here, if there is no signal you cannot print the signal

**Comment by louis-bompart** [comment]
New feature: You can now hook your own console/logger into the stdout and stderr of the SFDX process.
The default behaviour is no 'live logger'.

File: src/core/commandRunner.ts [hunk 2]
```diff
@@ -1,41 +1,53 @@
...
+          reject(stderr)
         } else {
-          message = e
+          resolve(stdout)
         }
+      })
+      if (this.useLiveLog) {
+        execProcess.stdout.on("data", data => {
+          this.customLogger.info(`${data}`)
+        })
```
Comment (erocheleau) [inline src/core/commandRunner.ts]: Missing `;`

File: src/core/commandRunner.ts [hunk 3]
```diff
@@ -1,41 +1,53 @@
...
+
+        execProcess.stderr.on("data", data => {
+          this.customLogger.error(`${data}`)
+        })
```
Comment (erocheleau) [inline src/core/commandRunner.ts]: `;`

File: src/core/commandRunner.ts [hunk 4]
```diff
@@ -1,41 +1,53 @@
...
 
-        reject(message)
+        execProcess.on("close", code => {
+          this.customLogger.info(`${fullCommand} exited with code ${code}`)
+        })
```
Comment (erocheleau) [inline src/core/commandRunner.ts]: `;`

File: src/core/commandRunner.ts [hunk 5]
```diff
@@ -1,41 +1,53 @@
...
-        let buffer = execSync(fullCommand, actualOptions)
-        resolve((buffer as Buffer).toString("utf8"))
-      } catch (e) {
-        let message: string
-        if (e !== undefined && e.stderr !== undefined) {
-          message = e["stderr"].toString("utf8")
+      let execProcess = exec(fullCommand, actualOptions, (error, stdout, stderr) => {
+        const errorMessage = stderr.toString()
+        if (error || errorMessage !== "") {
+          reject(stderr)
```
Comment (erocheleau) [inline src/core/commandRunner.ts]: `;`

File: src/core/commandRunner.ts [hunk 6]
```diff
@@ -1,41 +1,53 @@
...
         } else {
-          message = e
+          resolve(stdout)
```
Comment (erocheleau) [inline src/core/commandRunner.ts]: `;`

File: src/core/commandRunner.ts [hunk 7]
```diff
@@ -1,41 +1,53 @@
...
 
-      try {
-        let buffer = execSync(fullCommand, actualOptions)
-        resolve((buffer as Buffer).toString("utf8"))
-      } catch (e) {
-        let message: string
-        if (e !== undefined && e.stderr !== undefined) {
-          message = e["stderr"].toString("utf8")
+      let execProcess = exec(fullCommand, actualOptions, (error, stdout, stderr) => {
+        const errorMessage = stderr.toString()
```
Comment (erocheleau) [inline src/core/commandRunner.ts]: `;`

Comment (louis-bompart) [reply src/core/commandRunner.ts]: 👍 I'll add enforce the semicolon rule on tslint to fix that

**Comment by marcantoineveilleux** [comment]
Did you check very large output commands? Like `sfdx force:doc:commands:list --json`?

File: src/core/commandRunner.ts [hunk 8]
```diff
@@ -1,59 +1,71 @@
...
+      let execProcess = exec(fullCommand, actualOptions, (error, stdout, stderr) => {
+        const errorMessage = stderr.toString();
+        if (error || errorMessage !== "") {
+          reject(stderr);
         } else {
-          message = e
+          resolve(stdout);
         }
+      });
+      if (this.useLiveLog) {
```
[the remark this answers is not in the document]
Comment (erocheleau) [reply src/core/commandRunner.ts]: Could wrap those event bindings in another function for code clarity. But not necessarily required.

**Comment by louis-bompart** [comment]
@marcantoineveilleux I did and it worked.
@erocheleau I will remove the linting changes but I'll ignore the semicolon comment that you reported because I can't do both (because prettier run as a commit task to clean that).

File: src/core/commandRunner.ts (hunk 7 above)
Comment (marcantoineveilleux) [reply src/core/commandRunner.ts]: Pourquoi aller à l'encontre du linter? Pourquoi pas mettre des ; partout?

File: package.json [hunk 9]
```diff
@@ -46,7 +46,7 @@
   },
   "lint-staged": {
     "{src,test}/**/*.ts": [
-      "prettier --write",
+      "prettier --write --no-semi",
```
Comment (marcantoineveilleux) [inline package.json]: je crois que c'est l'inverse, on devrait les avoir partout et utiliser le plus possible les options OOTB de prettier.

Comment (louis-bompart) [reply package.json]: Well: @erocheleau said:
>Please revert all the changes that are not in the files you are changing on purpose.
> If you want to update the linter, it is very valid, but please make it a different PR.

File: src/core/commandRunner.ts [hunk 10]
```diff
@@ -1,43 +1,64 @@
...
-      try {
-        let buffer = execSync(fullCommand, actualOptions);
-        resolve((buffer as Buffer).toString("utf8"));
-      } catch (e) {
-        let message: string;
-        if (e !== undefined && e.stderr !== undefined) {
-          message = e["stderr"].toString("utf8");
+      // Execute the command. If there's an error thrown or the stderr is not empty, it will reject the promise.
+      // Otherwise, it will resolve with the content of the stdout.
+      let execProcess = exec(fullCommand, actualOptions, (error, stdout, stderr) => {
```
Comment (marcantoineveilleux) [inline src/core/commandRunner.ts]: ça, dans mes test ya longtemps, ça faisait en sorte que les commandes avec bcp d'output marchait pas. genre sfdx:doc:display. Peux-tu t'assurer que ça marche maintenant?

File: src/core/commandRunner.ts [hunk 11]
```diff
@@ -1,43 +1,64 @@
...
+}
+
+export interface CustomLogger {
+  error(data: string): void;
+  info(data: string): void;
 }
 
 export class CommandRunner implements ICommandRunner {
-  constructor(private SFDXPath: string) {}
+  constructor(
```
Comment (marcantoineveilleux) [inline src/core/commandRunner.ts]: J'crois qu'on devrait faire un object d'options avec des params par défault.

File: src/core/commandRunner.ts (hunk 10 above)
Comment (louis-bompart) [reply src/core/commandRunner.ts]: Yep ca marche

File: src/core/commandRunner.ts (hunk 11 above)
Comment (louis-bompart) [reply src/core/commandRunner.ts]: Done!

**Pull request merged**